File: CURRENT/java/src/org/mariuszgromada/math/mxparser/parsertokens/Operator.java
/*
 * @(#)Operator.java        5.0.0    2022-03-20
 *
 * You may use this software under the condition of "Simplified BSD License"
 *
 * Copyright 2010-2022 MARIUSZ GROMADA. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY <MARIUSZ GROMADA> ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of MARIUSZ GROMADA.
 *
 * If you have any questions/bugs feel free to contact:
 *
 *     Mariusz Gromada
 *     mariuszgromada.org@gmail.com
 *     http://mathparser.org
 *     http://mathspace.pl
 *     http://janetsudoku.mariuszgromada.org
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://mariuszgromada.github.io/MathParser.org-mXparser
 *     http://mxparser.sourceforge.net
 *     http://bitbucket.org/mariuszgromada/mxparser
 *     http://mxparser.codeplex.com
 *     http://github.com/mariuszgromada/Janet-Sudoku
 *     http://janetsudoku.codeplex.com
 *     http://sourceforge.net/projects/janetsudoku
 *     http://bitbucket.org/mariuszgromada/janet-sudoku
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://scalarmath.org/
 *     https://play.google.com/store/apps/details?id=org.mathparser.scalar.lite
 *     https://play.google.com/store/apps/details?id=org.mathparser.scalar.pro
 *
 *                              Asked if he believes in one God, a mathematician answered:
 *                              "Yes, up to isomorphism."
 */
package org.mariuszgromada.math.mxparser.parsertokens;

import org.mariuszgromada.math.mxparser.mXparser;
/**
 * Operators - mXparser tokens definition.
 *
 * @author         <b>Mariusz Gromada</b><br>
 *                 <a href="mailto:mariuszgromada.org@gmail.com">mariuszgromada.org@gmail.com</a><br>
 *                 <a href="http://mathspace.pl" target="_blank">MathSpace.pl</a><br>
 *                 <a href="http://mathparser.org" target="_blank">MathParser.org - mXparser project page</a><br>
 *                 <a href="http://github.com/mariuszgromada/MathParser.org-mXparser" target="_blank">mXparser on GitHub</a><br>
 *                 <a href="http://mxparser.sourceforge.net" target="_blank">mXparser on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/mxparser" target="_blank">mXparser on Bitbucket</a><br>
 *                 <a href="http://mxparser.codeplex.com" target="_blank">mXparser on CodePlex</a><br>
 *                 <a href="http://janetsudoku.mariuszgromada.org" target="_blank">Janet Sudoku - project web page</a><br>
 *                 <a href="http://github.com/mariuszgromada/Janet-Sudoku" target="_blank">Janet Sudoku on GitHub</a><br>
 *                 <a href="http://janetsudoku.codeplex.com" target="_blank">Janet Sudoku on CodePlex</a><br>
 *                 <a href="http://sourceforge.net/projects/janetsudoku" target="_blank">Janet Sudoku on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/janet-sudoku" target="_blank">Janet Sudoku on BitBucket</a><br>
 *                 <a href="https://play.google.com/store/apps/details?id=org.mathparser.scalar.lite" target="_blank">Scalar Free</a><br>
 *                 <a href="https://play.google.com/store/apps/details?id=org.mathparser.scalar.pro" target="_blank">Scalar Pro</a><br>
 *                 <a href="http://scalarmath.org/" target="_blank">ScalarMath.org</a><br>
 *
 * @version        5.0.0
 */
public final class Operator {
	/*
	 * Operator - token type id.
	 */
	public static final int TYPE_ID 				= 1;
	public static final String TYPE_DESC			= "Operator";
	/*
	 * Operator - tokens id.
	 */
	public static final int PLUS_ID 				= 1;
	public static final int MINUS_ID 				= 2;
	public static final int MULTIPLY_ID 			= 3;
	public static final int DIVIDE_ID 				= 4;
	public static final int POWER_ID 				= 5;
	public static final int FACT_ID					= 6;
	public static final int MOD_ID					= 7;
	public static final int PERC_ID					= 8;
	public static final int TETRATION_ID			= 9;
	public static final int SQUARE_ROOT_ID			= 10;
	public static final int CUBE_ROOT_ID			= 11;
	public static final int FOURTH_ROOT_ID			= 12;
	/*
	 * Operator - tokens key words.
	 */
	public static final String PLUS_STR 			= "+";
	public static final String MINUS_STR 			= "-";
	public static final String MULTIPLY_STR 		= "*";
	public static final String MULTIPLY_STR_UNI_1	= "×";
	public static final String MULTIPLY_STR_UNI_2	= "⨉";
	public static final String MULTIPLY_STR_UNI_3	= "∙";
	public static final String DIVIDE_STR 			= "/";
	public static final String DIVIDE_STR_UNI_1		= "÷";
	public static final String POWER_STR 			= "^";
	public static final String FACT_STR 			= "!";
	public static final String MOD_STR				= "#";
	public static final String PERC_STR				= "%";
	public static final String TETRATION_STR		= "^^";
	public static final String SQUARE_ROOT_STR		= "√";
	public static final String CUBE_ROOT_STR		= "∛";
	public static final String FOURTH_ROOT_STR		= "∜";
	/*
	 * Operator - syntax.
	 */
	public static final String PLUS_SYN 			= SyntaxStringBuilder.binaryOperator(PLUS_STR);
	public static final String MINUS_SYN 			= SyntaxStringBuilder.binaryOperator(MINUS_STR);
	public static final String MULTIPLY_SYN 		= SyntaxStringBuilder.binaryOperator(MULTIPLY_STR);
	public static final String MULTIPLY_SYN_UNI_1	= SyntaxStringBuilder.binaryOperator(MULTIPLY_STR_UNI_1);
	public static final String MULTIPLY_SYN_UNI_2	= SyntaxStringBuilder.binaryOperator(MULTIPLY_STR_UNI_2);
	public static final String MULTIPLY_SYN_UNI_3	= SyntaxStringBuilder.binaryOperator(MULTIPLY_STR_UNI_3);
	public static final String DIVIDE_SYN 			= SyntaxStringBuilder.binaryOperator(DIVIDE_STR);
	public static final String DIVIDE_SYN_UNI_1		= SyntaxStringBuilder.binaryOperator(DIVIDE_STR_UNI_1);
	public static final String POWER_SYN 			= SyntaxStringBuilder.binaryOperatorNoSpace(POWER_STR);
	public static final String FACT_SYN 			= SyntaxStringBuilder.unaryRightOperator(FACT_STR);
	public static final String MOD_SYN				= SyntaxStringBuilder.binaryOperator(MOD_STR);
	public static final String PERC_SYN				= SyntaxStringBuilder.unaryRightOperator(PERC_STR);
	public static final String TETRATION_SYN		= SyntaxStringBuilder.binaryOperatorNoSpace(TETRATION_STR);
	public static final String SQUARE_ROOT_SYN		= SyntaxStringBuilder.unaryLeftOperator(SQUARE_ROOT_STR);
	public static final String CUBE_ROOT_SYN		= SyntaxStringBuilder.unaryLeftOperator(CUBE_ROOT_STR);
	public static final String FOURTH_ROOT_SYN		= SyntaxStringBuilder.unaryLeftOperator(FOURTH_ROOT_STR);
	/*
	 * Operator - tokens description.
	 */
	public static final String PLUS_DESC 			= "Addition";
	public static final String MINUS_DESC 			= "Subtraction";
	public static final String MULTIPLY_DESC 		= "Multiplication";
	public static final String DIVIDE_DESC 			= "Division";
	public static final String POWER_DESC 			= "Exponentiation";
	public static final String FACT_DESC			= "Factorial";
	public static final String MOD_DESC				= "Modulo function";
	public static final String PERC_DESC			= "Percentage";
	public static final String TETRATION_DESC		= "Tetration (hyper-4, power tower, exponential tower)";
	public static final String SQUARE_ROOT_DESC		= "Square root function represented as unary left operator";
	public static final String CUBE_ROOT_DESC		= "Cube root function represented as unary left operator";
	public static final String FOURTH_ROOT_DESC		= "Fourth root function represented as unary left operator";
	/*
	 * Operator - since.
	 */
	public static final String PLUS_SINCE 			= mXparser.NAMEv10;
	public static final String MINUS_SINCE 			= mXparser.NAMEv10;
	public static final String MULTIPLY_SINCE 		= mXparser.NAMEv10;
	public static final String MULTIPLY_SINCE_UNI_1	= mXparser.NAMEv50;
	public static final String MULTIPLY_SINCE_UNI_2	= mXparser.NAMEv50;
	public static final String MULTIPLY_SINCE_UNI_3	= mXparser.NAMEv50;
	public static final String DIVIDE_SINCE 		= mXparser.NAMEv10;
	public static final String DIVIDE_SINCE_UNI_1	= mXparser.NAMEv50;
	public static final String POWER_SINCE 			= mXparser.NAMEv10;
	public static final String FACT_SINCE			= mXparser.NAMEv10;
	public static final String MOD_SINCE			= mXparser.NAMEv10;
	public static final String PERC_SINCE			= mXparser.NAMEv41;
	public static final String TETRATION_SINCE		= mXparser.NAMEv42;
	public static final String SQUARE_ROOT_SINCE	= mXparser.NAMEv50;
	public static final String CUBE_ROOT_SINCE		= mXparser.NAMEv50;
	public static final String FOURTH_ROOT_SINCE	= mXparser.NAMEv50;
}

File: CURRENT/typescript/org/mariuszgromada/math/mxparser/parsertokens/Operator.ts
import { mXparserConstants } from '../mXparserConstants';

/**
 * Operators - mXparserConstants tokens definition.
 * 
 * @author         <b>Mariusz Gromada</b><br>
 * <a href="mailto:mariuszgromada.org@gmail.com">mariuszgromada.org@gmail.com</a><br>
 * <a href="http://github.com/mariuszgromada/MathParser.org-mXparser" target="_blank">mXparser on GitHub</a><br>
 * 
 * @version        4.2.0
 * @class
 */
export class Operator {
    public static TYPE_ID: number = 1;

    public static TYPE_DESC: string = "Operator";

    public static PLUS_ID: number = 1;

    public static MINUS_ID: number = 2;

    public static MULTIPLY_ID: number = 3;

    public static DIVIDE_ID: number = 4;

    public static POWER_ID: number = 5;

    public static FACT_ID: number = 6;

    public static MOD_ID: number = 7;

    public static PERC_ID: number = 8;

    public static TETRATION_ID: number = 9;

    public static PLUS_STR: string = "+";

    public static MINUS_STR: string = "-";

    public static MULTIPLY_STR: string = "*";

    public static DIVIDE_STR: string = "/";

    public static POWER_STR: string = "^";

    public static FACT_STR: string = "!";

    public static MOD_STR: string = "#";

    public static PERC_STR: string = "%";

    public static TETRATION_STR: string = "^^";

    public static PLUS_SYN: string = "a + b";

    public static MINUS_SYN: string = "a - b";

    public static MULTIPLY_SYN: string = "a * b";

    public static DIVIDE_SYN: string = "a / b";

    public static POWER_SYN: string = "a^b";

    public static FACT_SYN: string = "n!";

    public static MOD_SYN: string = "a # b";

    public static PERC_SYN: string = "n%";

    public static TETRATION_SYN: string = "a^^n";

    public static PLUS_DESC: string = "Addition";

    public static MINUS_DESC: string = "Subtraction";

    public static MULTIPLY_DESC: string = "Nultiplication";

    public static DIVIDE_DESC: string = "Division";

    public static POWER_DESC: string = "Exponentiation";

    public static FACT_DESC: string = "Factorial";

    public static MOD_DESC: string = "Modulo function";

    public static PERC_DESC: string = "Percentage";

    public static TETRATION_DESC: string = "Tetration (hyper-4, power tower, exponential tower)";

    public static PLUS_SINCE: string; public static PLUS_SINCE_$LI$(): string { if (Operator.PLUS_SINCE == null) { Operator.PLUS_SINCE = mXparserConstants.NAMEv10; }  return Operator.PLUS_SINCE; }

    public static MINUS_SINCE: string; public static MINUS_SINCE_$LI$(): string { if (Operator.MINUS_SINCE == null) { Operator.MINUS_SINCE = mXparserConstants.NAMEv10; }  return Operator.MINUS_SINCE; }

    public static MULTIPLY_SINCE: string; public static MULTIPLY_SINCE_$LI$(): string { if (Operator.MULTIPLY_SINCE == null) { Operator.MULTIPLY_SINCE = mXparserConstants.NAMEv10; }  return Operator.MULTIPLY_SINCE; }

    public static DIVIDE_SINCE: string; public static DIVIDE_SINCE_$LI$(): string { if (Operator.DIVIDE_SINCE == null) { Operator.DIVIDE_SINCE = mXparserConstants.NAMEv10; }  return Operator.DIVIDE_SINCE; }

    public static POWER_SINCE: string; public static POWER_SINCE_$LI$(): string { if (Operator.POWER_SINCE == null) { Operator.POWER_SINCE = mXparserConstants.NAMEv10; }  return Operator.POWER_SINCE; }

    public static FACT_SINCE: string; public static FACT_SINCE_$LI$(): string { if (Operator.FACT_SINCE == null) { Operator.FACT_SINCE = mXparserConstants.NAMEv10; }  return Operator.FACT_SINCE; }

    public static MOD_SINCE: string; public static MOD_SINCE_$LI$(): string { if (Operator.MOD_SINCE == null) { Operator.MOD_SINCE = mXparserConstants.NAMEv10; }  return Operator.MOD_SINCE; }

    public static PERC_SINCE: string; public static PERC_SINCE_$LI$(): string { if (Operator.PERC_SINCE == null) { Operator.PERC_SINCE = mXparserConstants.NAMEv41; }  return Operator.PERC_SINCE; }

    public static TETRATION_SINCE: string; public static TETRATION_SINCE_$LI$(): string { if (Operator.TETRATION_SINCE == null) { Operator.TETRATION_SINCE = mXparserConstants.NAMEv42; }  return Operator.TETRATION_SINCE; }
}
Operator["__class"] = "org.mariuszgromada.math.mxparser.parsertokens.Operator";